src/util/expr_util.cpp

```diff
@@ -24,6 +24,8 @@ bool is_lvalue(const exprt &expr)
     return is_lvalue(to_index_expr(expr).array());
   else if(expr.id() == ID_member)
     return is_lvalue(to_member_expr(expr).compound());
+  else if(expr.id() == ID_ptrmember)
+    return true;
```

Comment:
@tautschnig @kroening We are calling this function to check that the targets specified in assigns clause for functions are valid l-values. The `ptrmember` targets there aren't decomposed to deref + member before type checking. Should we be calling some utility function for the decomposition, or should we add this case here?

If we add this case, it should do something similar to the `member` case and not return `true` probably, because `a->b` == `(*a).b`.

Reply:
@peterschrammel @martin-cs @chrisr-diffblue any thoughts on this too?

Reply:
There is something wrong here. ID_ptrmember is an expression that is internal to the C front-end, and should not be visible outside of it. It's possible that `is_value` is internal to the C front-end as well, and should possibly move there.

Reply:
Thanks.
It looks like `is_lvalue` is supposed to be used only on goto expressions, not C expressions. So calling it from out instrumentation code is okay, but calling it from our compilation code is not. 
Is there a similar function to check if an expression is a valid C l-value?

Reply:
Over in another slack thread Daniel mentioned "Within the C frontend, you can tell whether or not an expression is an lvalue by querying expr.get_bool(ID_C_lvalue)"

Reply:
For `__CPROVER_object(arg)` calls in the assigns clause, we want to make a special case and want to type check that the `deref(arg)` is an lvalue.

Reply:
After a few offline discussions, CBMC now evaluates for `lvalue` expressions in assigns clauses after type checking using `get_bool(ID_C_lvalue)`.